go/client/batch_retries_test.go

```diff
@@ -0,0 +1,378 @@
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	regrpc "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	repb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	spb "google.golang.org/genproto/googleapis/rpc/status"
+)
+
+type flakyBatchServer struct {
+	numErrors      int // A counter of errors the server has returned thus far.
+	updateRequests []*repb.BatchUpdateBlobsRequest
+	readRequests   []*repb.BatchReadBlobsRequest
+}
+
+func (f *flakyBatchServer) FindMissingBlobs(ctx context.Context, req *repb.FindMissingBlobsRequest) (*repb.FindMissingBlobsResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+func (f *flakyBatchServer) BatchReadBlobs(ctx context.Context, req *repb.BatchReadBlobsRequest) (*repb.BatchReadBlobsResponse, error) {
+	f.readRequests = append(f.readRequests, req)
+	if f.numErrors < 1 {
+		f.numErrors++
+		resp := &repb.BatchReadBlobsResponse{
+			Responses: []*repb.BatchReadBlobsResponse_Response{
+				{Digest: digest.TestNew("a", 1).ToProto(), Status: &spb.Status{Code: int32(codes.OK)}, Data: []byte{1}},
+				// all retriable errors.
+				{Digest: digest.TestNew("b", 1).ToProto(), Status: &spb.Status{Code: int32(codes.Internal)}},
+				{Digest: digest.TestNew("c", 1).ToProto(), Status: &spb.Status{Code: int32(codes.Canceled)}},
+				{Digest: digest.TestNew("d", 1).ToProto(), Status: &spb.Status{Code: int32(codes.Aborted)}},
+			},
+		}
+		return resp, nil
+	}
+	if f.numErrors < 2 {
+		f.numErrors++
+		resp := &repb.BatchReadBlobsResponse{
+			Responses: []*repb.BatchReadBlobsResponse_Response{
+				{Digest: digest.TestNew("b", 1).ToProto(), Status: &spb.Status{Code: int32(codes.OK)}, Data: []byte{2}},
+				// all retriable errors.
+				{Digest: digest.TestNew("c", 1).ToProto(), Status: &spb.Status{Code: int32(codes.Internal)}},
+				{Digest: digest.TestNew("d", 1).ToProto(), Status: &spb.Status{Code: int32(codes.Canceled)}},
+			},
+		}
+		return resp, nil
+	}
+	if f.numErrors < 3 {
+		f.numErrors++
+		resp := &repb.BatchReadBlobsResponse{
+			Responses: []*repb.BatchReadBlobsResponse_Response{
+				// One non-retriable error.
+				{Digest: digest.TestNew("c", 1).ToProto(), Status: &spb.Status{Code: int32(codes.Internal)}},
+				{Digest: digest.TestNew("d", 1).ToProto(), Status: &spb.Status{Code: int32(codes.PermissionDenied)}},
+			},
+		}
+		return resp, nil
+	}
+	// Will not be reached.
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+func (f *flakyBatchServer) GetTree(req *repb.GetTreeRequest, stream regrpc.ContentAddressableStorage_GetTreeServer) error {
+	return status.Error(codes.Unimplemented, "")
+}
+
+func (f *flakyBatchServer) BatchUpdateBlobs(ctx context.Context, req *repb.BatchUpdateBlobsRequest) (*repb.BatchUpdateBlobsResponse, error) {
+	f.updateRequests = append(f.updateRequests, req)
+	if f.numErrors < 1 {
+		f.numErrors++
```

Comment:
Can you just increment the number of errors once at the beginning of the function?

Reply:
Yep.